primitives/src/pcs/mod.rs

```diff
@@ -0,0 +1,155 @@
+
+//! Polynomial Commitment Scheme
+pub mod errors;
+mod multilinear_kzg;
+pub mod prelude;
+mod structs;
+mod transcript;
+mod univariate_kzg;
+
+use ark_ec::PairingEngine;
+use ark_serialize::CanonicalSerialize;
+use ark_std::{rand::RngCore, vec::Vec};
+use errors::PCSError;
+
+/// This trait defines APIs for polynomial commitment schemes.
+/// Note that for our usage of PCS, we do not require the hiding property.
+pub trait PolynomialCommitmentScheme<E: PairingEngine> {
+    /// Prover parameters
+    type ProverParam;
+    /// Verifier parameters
+    type VerifierParam;
+    /// Structured reference string
+    type SRS;
+    /// Polynomial and its associated types
+    type Polynomial;
+    /// Polynomial input domain
+    type Point;
+    /// Polynomial Evaluation
+    type Evaluation;
+    /// Commitments
+    type Commitment: CanonicalSerialize;
+    /// Batch commitments
+    type BatchCommitment: CanonicalSerialize;
+    /// Proofs
+    type Proof;
+    /// Batch proofs
+    type BatchProof;
+
+    /// Build SRS for testing.
+    ///
+    /// - For univariate polynomials, `supported_size` is the maximum degree.
+    /// - For multilinear polynomials, `supported_size` is the number of
+    ///   variables.
+    ///
+    /// WARNING: THIS FUNCTION IS FOR TESTING PURPOSE ONLY.
+    /// THE OUTPUT SRS SHOULD NOT BE USED IN PRODUCTION.
+    fn gen_srs_for_testing<R: RngCore>(
+        rng: &mut R,
+        supported_size: usize,
+    ) -> Result<Self::SRS, PCSError>;
+
+    /// Trim the universal parameters to specialize the public parameters.
+    /// Input both `supported_degree` for univariate and
+    /// `supported_num_vars` for multilinear.
+    fn trim(
+        srs: &Self::SRS,
+        supported_degree: usize,
+        supported_num_vars: Option<usize>,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), PCSError>;
+
+    /// Generate a commitment for a polynomial
+    fn commit(
+        prover_param: &Self::ProverParam,
+        poly: &Self::Polynomial,
+    ) -> Result<Self::Commitment, PCSError>;
+
+    /// Generate a commitment for a list of polynomials
+    fn multi_commit(
+        prover_param: &Self::ProverParam,
+        polys: &[Self::Polynomial],
+    ) -> Result<Self::BatchCommitment, PCSError>;
+
+    /// On input a polynomial `p` and a point `point`, outputs a proof for the
+    /// same.
+    fn open(
+        prover_param: &Self::ProverParam,
+        polynomial: &Self::Polynomial,
+        point: &Self::Point,
+    ) -> Result<(Self::Proof, Self::Evaluation), PCSError>;
+
+    /// Input a list of MLEs, and a same number of points, and a transcript,
+    /// compute a multi-opening for all the polynomials.
+    fn multi_open(
+        prover_param: &Self::ProverParam,
+        multi_commitment: &Self::Commitment,
+        polynomials: &[Self::Polynomial],
+        points: &[Self::Point],
+    ) -> Result<(Self::BatchProof, Vec<Self::Evaluation>), PCSError>;
```

Comment:
1. what is `MLEs` in the code comments?
2. what is a `multi_commitment`? should explain in comment, and explain
   - what API is used to generate such a `multi_comm`
   - how is different from `Self::BatchCommitment`
3. why `multi_open(multi_commitment: &Self::Commitment)`, but `batch_verify(multi_commitment: &Self::BatchCommitment)` ?